### DISCUSSION TITLE
Generals vs Combination setter

### DISCUSSION THREAD

**Discussion by Jean-Gabriel1982** [GrandOrgue/grandorgue#1908] [General]
Hello, I would like to implement a digital visual of the state of my Generals in my main panel on GrandOrgue.
Thanks to @eturpault, whom I thank for his call yesterday and whom I salute in this space, I was able to graphically integrate the elements I wanted.

Only... I can't get it to work! 😄 

I have 12 physical buttons under my keyboard block. 10 buttons are configured as generals that I save (not fixed combination!) the other two are used to navigate banks A to Z (the “generals” panel of GrandOrgue).

On this generals panel, when I press my physical buttons, we can clearly see that the Midi signal is working. Only, on my panel, nothing changes. 
Bank A stays on A, while it advances (B, C, D etc.) in the “generals” panel. The same for the combinations which remain at 000… could you help me?

**Comment by eturpault** [comment]
Hello @Jean-Gabriel1982 , in your panel, have you actually defined a PanelElement section with type GeneralLabel (and SequencerLabel for the combinations) ? It should show the same bank letter as in the Generals panel (and same combination number as in the Combination Setter panel).

**Reply by Jean-Gabriel1982** [reply]
Hello @eturpault yes I defined that .. well I don't understand why but now the letter of panel "Generals" works .. 

The SequencerLabel does not work. 
Maybe I misconfigured it ?
I saved my 10 buttons in the Generals Panel (so I saved 10 of the 50 buttons in this panel) .. maybe it is not here but in the combination setter panel ? 
I don't really understand how this panel works ..

**Comment by eturpault** [comment]
I had a call with @Jean-Gabriel1982. What would fit his need (and probably the need of other users) would be to add the following setters (with Type values that I am proposing) :

- **GeneralNumberLabel** : it would show the number of the general combination which has been activated by pressing on a setter from General01 to General50. If GC is pressed, this label would become empty.
- **GeneralNumberPrev** / **GeneralNumberNext** : it would permit to select and apply the previous / next general combination if a general combination is currently selected. If no general is currently selected, it would select/apply the General50 (if Prev) or General01 (if Next). It would be like a sequencer between the General combinations. The GeneralNumberLabel would then show the number of the newly selected General.

@larspalo @oleg68 : could such 3 setters be added to GO ?

**Reply by oleg68** [reply]
They already exist on the Setter panel

**Reply by eturpault** [reply]
If I am not wrong in the Setter panel the buttons are acting on the sequencer only. The request here is to act on the General combinations, so to have in the Generals panel two new buttons (Previous general, Next general) and one new label (showing the number of the applied general, not of the selected general bank).

**Reply by larspalo** [reply]
@eturpault It seems that there are some confusion about how the (Setter) Generals are supposed to behave in contrast to how the (Setter) Sequencer combinations are supposed to work. They are two separate ways of how combination systems are supposed to work - not to be mixed up!

The Generals already gives the feedback by means of what General piston is pushed in combination with what Bank is active. That's how the English/American combination system is working. In other words - they want the feedback in form of what button is actually "lit" or "depressed" on the console when that combination is active. Also you have the GeneralLabel indicating what Bank you're on. In the english/american system you never step to "the next" (or previous) general - you push the button you want to "jump" to immediately.

In contrast to that, the "modern european", "free combination" system, doesn't have physical buttons that lit up or is depressed in the same sense as the english/american system. One usually have just the feedback on a display what combination number you're on and then step to next/back from there, or by number entry would "jump" to somewhere else. But generally speaking you're supposed to sequentially step through a series of combinations. That's how it's supposed to work.

So, reading the discussion above it seems that you should be investigating the setter Sequencer and not the Generals to get the desired behaviour.

**Comment by larspalo** [comment]
@Jean-Gabriel1982 @eturpault On the other hand, the Setter Sequencer _can_ be used as a kind of banked generals system with the numbering feature and the next/previous if one so desires.

The "buttons" to implement then would be the (L0, L1, L2, L3, L4, L5, L6, L7, L8, L9) for direct access, the next/previous and then the SequencerLabel to display the number.

Someone might also want the +1,-1, +10, -10, +100, -100 and 000 buttons for advancing more quickly through the "banks".

So, the Sequencer _can_ be used more in the english/american style too if one really desires it.

**Comment by eturpault** [comment]
Thanks for the explanations.
So @Jean-Gabriel1982 what you want to implement has to use the Sequencer (Combination Setter panel) and not the Generals.
What you presented about your configuration 
> I have 12 physical buttons under my keyboard block. 10 buttons are configured as generals that I save (not fixed combination!) the other two are used to navigate banks A to Z (the “generals” panel of GrandOrgue).

should be rather set by assigning your 10 physical buttons to the buttons __0, __1, ... __9 of the Sequencer, and the two remaining buttons assigned to the Previous and Next buttons of the Sequencer. You will need also the possibility to press on the Set and Current buttons to memorize combinations.

**Reply by Jean-Gabriel1982** [reply]
Thank you very much everyone for your clarifications! @eturpault, the problem is that by configuring my 10 buttons with __0, __1, etc, and the two remaining buttons as previous and next.. I would only have the possibility of having 10 combinations which is not what not? because I couldn't register two different combinations on the same button?! 

I understood @larspalo's explanations, following my discussion with @ericturpault for the difference between Generals and Combination Setter, it's much clearer now in my mind... 

But let me insist a little :- ). it would be interesting to have the possibility to "mix" the two (Generals and Combination setter), in the sense that the user could choose if he wants +10 or +100 buttons or a letter (or number) representing a bank of combinations. 

For my part, I would prefer the second solution. But my personal opinion does not count, rather I would argue that everyone has a console arranged differently with more or fewer buttons and giving the choice could be a "plus" ..

**Reply by larspalo** [reply]
@Jean-Gabriel1982 From the last combination L9 you can use Next to get to a new "bank", or from a higher L0 you can use Previous to get to a lower "bank". The +/- 10 or 100 also can jump to different "banks". This still retains the stepping possibility. With a SequencerLabel you always know where you are.

For your other suggestion I have this story:

A long time ago, when I started playing pipe organs (around 1987), and the digital/electric combination systems were much more primitive than they are today, there was indeed a kind of "Letter + Number" identification when the number of combinations would be above a certain number. If I remember correctly it was at that time marketed by a German organ parts supplier named Laukhuff - now out of business.

In that particular system you'd have number buttons from 1 to 10 and letter buttons from A to J to give you 100 free combinations. They would be organized in a series that could be stepped through forward and backward and also directly accessed by first pushing the letter and then the number, or within each "letter bank" the numbers only would directly activate that combination. To extend that up to 200 free combinations I think there was a star button available (I encountered this around 1991 in Luleå Cathedral).

With later extensions of that system you could have A 1 to J 10 (100 combinations) and then 1 A 1 to 1 J 10 (next 100 combinations) and so on up to perhaps 9 A 1 to 9 J 10 (which would then total 1000 free combinations). Further expansions to that system was eventually made until that company was out of business somewhere in 2021 or so.

Adding yet another possible way of organizing, or mapping, the Sequencer or Combination Setter of GO with its 0 - 999 (1000 combinations) would certainly be possible to do, but is it really worth it? If someone thinks it is, I'd really appreciate if they also learned to do the programming necessary to achieve it.

**Reply by Jean-Gabriel1982** [reply]
Hello, @larspalo thank you for all these explanations. 

So I built my combiner with the SequencerLabel. It works quite well :-) @ericturpault, ultimately no need for the Current button. Just Set and press the button corresponding to the combination you want to save. 

Only downside, I have 10 buttons and the last possible number in the sequencer is __9. 
So I assigned my button 10 to the __0 function, which resets the sequencer to 000. My buttons 11 and 12 are assigned to +10 and -10.
